include/librealsense2/dds/dds-topic-reader.cpp

```diff
@@ -0,0 +1,109 @@
+#include "dds-participant.h"
+#include "dds-utilities.h"
+
+#include <fastdds/dds/domain/DomainParticipant.hpp>
+#include <fastdds/dds/subscriber/Subscriber.hpp>
+#include <fastdds/dds/subscriber/DataReader.hpp>
+#include <fastdds/dds/topic/Topic.hpp>
+#include <fastdds/dds/core/status/SubscriptionMatchedStatus.hpp>
+
+
+namespace librealsense {
+namespace dds {
+
+
+dds_topic_reader::dds_topic_reader( std::shared_ptr< dds_topic > const & topic )
+    : _topic( topic )
+{
+}
+
+
+dds_topic_reader::~dds_topic_reader()
+{
+    if( _subscriber )
+    {
+        if( _reader )
+            DDS_API_CALL_NO_THROW( _subscriber->delete_datareader( _reader ) );
+        if( _topic->get_participant()->is_valid() )
+            DDS_API_CALL_NO_THROW( _topic->get_participant()->get()->delete_subscriber( _subscriber ) );
+    }
+}
+
+
+dds_topic_reader::reader_qos::reader_qos()
+    : super( eprosima::fastdds::dds::DATAREADER_QOS_DEFAULT )
+{
+    // The 'depth' parameter of the History defines how many samples are stored before starting to
+    // overwrite them with newer ones.
+    history().kind = eprosima::fastdds::dds::KEEP_LAST_HISTORY_QOS;
+    history().depth = 10;  // default is 1
+
+    // We don't want to miss connection/disconnection events
+    // (default is best-effort)
+    reliability().kind = eprosima::fastdds::dds::RELIABLE_RELIABILITY_QOS;
+
+    // The Subscriber receives samples from the moment it comes online, not before
+    // (this is the default, but bears repeating)
+    durability().kind = eprosima::fastdds::dds::VOLATILE_DURABILITY_QOS;
+
+    // The writer-reader handshake is always done on UDP, even when data_sharing (shared memory) is used for
+    // actual messaging. This means it is possible to send a message and receive it on the reader's
+    // side even before the UDP handshake is complete:
+    // 
+    //   1. The writer goes up and broadcasts its presence periodically; no readers exist
+    //   2. The reader joins and broadcasts its presence, again periodically; it doesn<92>t know about the writer yet
+    //   3. The writer sees the reader (in-between broadcasts) so sends a message
+    //   4. The reader gets the message and discards it because it does not yet recognize the writer
+    // 
+    // This depends on timing. When shared memory is on, step 3 is so fast that this miscommunication is much more
+    // likely. This is a known gap in the DDS standard.
+    //
+    // We can either insert a sleep between writer creation and message sending or we can disable
+    // data_sharing for this topic, which we did here.
+    // 
+    // See https://github.com/eProsima/Fast-DDS/issues/2641
+    //
+    data_sharing().off();
+}
+
+
+void dds_topic_reader::run( reader_qos const & rqos )
+{
+    // The Subscriber manages the activities of several DataReader entities
+    _subscriber = DDS_API_CALL(
+        _topic->get_participant()->get()->create_subscriber( eprosima::fastdds::dds::SUBSCRIBER_QOS_DEFAULT ) );
```

Comment:
As discussed, can have a single subscriber in the domain, created and managed by `dds_participant` class